### PR TITLE
[SPARK-10065][SQL] avoid the extra copy when generate unsafe array

### PR DESCRIPTION
**Pull request by cloud-fan**
The reason for this extra copy is that we iterate the array twice: calculate elements data size and copy elements to array buffer.

A simple solution is to follow `createCodeForStruct`, we can dynamically grow the buffer when needed and thus don't need to know the data size ahead.

This PR also include some typo and style fixes, and did some minor refactor to make sure `input.primitive` is always variable name not code when generate unsafe code.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @davis @viirya

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala [hunk 1]
```diff
@@ -393,22 +362,22 @@ object GenerateUnsafeProjection extends CodeGenerator[Seq[Expression], UnsafePro
...
+        GeneratedExpressionCode(code, isNull, fieldName)
       }
-      val converter = createCodeForStruct(ctx, tmp, fieldEvals, fieldTypes)
+      val converter = createCodeForStruct(ctx, input.primitive, fieldEvals, fieldTypes)
       val code = s"""
         ${input.code}
          UnsafeRow $output = null;
          final boolean $outputIsNull = ${input.isNull};
          if (!$outputIsNull) {
-           final InternalRow $tmp = ${input.primitive};
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala]: Why remove this? I think it's good to have that, `input.primitive` could be an expression

**Comment by davies** [comment]
@cloud-fan This change can only remove one copy for ArrayType and MapType, we still have one additional copy for each level of Array or Struct. Can we copy the elements into target UnsafeRow in one pass? (it will be harder, I think it worth the effort since the changes in this PR is also not trivial).

BTW, it's better to separate real change and refactors to reduce the burden of review, especially when the PR is large, thanks!

**Comment by davis** [comment]
@davies

**Comment by cloud-fan** [comment]
Hi @davies , I have moved the refactor part to https://github.com/apache/spark/pull/8613.

I'd like to open another PR to remove the one more additional copy for each level of Array or Struct, as this JIRA is only for array.

**Comment by davies** [comment]
LGTM, will merge it once you fix the conflict.

**Comment by davies** [comment]
Merged into master, thanks!

**Pull request closed without merging**